### PR TITLE
OCC-18: PriceVariantsPart breaks when you have a TextProductAttributeField with unrestricted values

### LINKED ISSUE #106: PriceVariantsPart breaks when you have a TextProductAttributeField with unrestricted values.
## Reproduction steps:
- Create a new content type.
- Add parts `ProductPart` and `PriceVariantsPart`.
- Add two `TextProductAttributeField` fields:
  - Size: has predefined values (e.g. Small, Medium, Large) and "Restrict to predefined values" is checked.
  - Color: don't change anything, i.e. predefined values is empty "Restrict to predefined values" is unchecked. (I just named this color but could be anything else, a plausible scenario is a t-shirt shop where you can add a custom text on your order.)
- Create a new content item.
- Set SKU as you wish, you should see 3 price variant fields for Small, Medium and Large prices.

## Problems:
- The currency dropdown stays as `---` instead of `USD` as configured (note, this is not related to [the other issue](https://github.com/OrchardCMS/OrchardCore.Commerce/issues/105), if I replace `PriceVariantsPart` with `PricePart` it initializes correctly)
- When I press Publish the values aren't saved.
  - When I try to add it to the cart anyway I get "Can't add product sample because it doesn't have a price."

before publish:
image
after publish:
image

During publish I see this error in the console log:
```
fail: OrchardCore.ContentManagement.Display.ContentItemDisplayCoordinator[0]
      IContentPartDisplayDriver thrown from OrchardCore.Commerce.Drivers.PriceVariantsPartDisplayDriver by RuntimeBinderException
      Microsoft.CSharp.RuntimeBinder.RuntimeBinderException: 'Newtonsoft.Json.Linq.JValue' does not contain a definition for 'RemoveAll'
         at CallSite.Target(Closure , CallSite , Object )
         at OrchardCore.Commerce.Drivers.PriceVariantsPartDisplayDriver.UpdateAsync(PriceVariantsPart part, IUpdateModel updater, UpdatePartEditorContext context) in E:\aux_c\repositories\orchard-core\OrchardCore.Commerce\OrchardCore.Commerce\Drivers\PriceVarian
tsPartDisplayDriver.cs:line 49
         at OrchardCore.ContentManagement.Display.ContentDisplay.ContentPartDisplayDriver`1.OrchardCore.ContentManagement.Display.ContentDisplay.IContentPartDisplayDriver.UpdateEditorAsync(ContentPart contentPart, ContentTypePartDefinition typePartDefinition, Up
dateEditorContext context)
         at OrchardCore.ContentManagement.Display.ContentItemDisplayCoordinator.<>c.<<UpdateEditorAsync>b__9_0>d.MoveNext()
      --- End of stack trace from previous location ---
         at OrchardCore.Modules.InvokeExtensions.InvokeAsync[TEvents,T1,T2,T3](IEnumerable`1 events, Func`5 dispatch, T1 arg1, T2 arg2, T3 arg3, ILogger logger)
```

## Notes

If I add predefined values and check the box for the other field it starts to work correctly. I think a solution would be to filter the product attributes the `PriceVariantsPart` handles, to only include those with "Restrict to predefined values".

### PR DESCRIPTION
**Pull request by Psichorex**
OCC-18
Fixes #106

### CODE REVIEW

File: src/Modules/OrchardCore.Commerce/Drivers/PriceVariantsPartDisplayDriver.cs [hunk 1]
```diff
@@ -50,7 +55,8 @@ public class PriceVariantsPartDisplayDriver : ContentPartDisplayDriver<PriceVari
                 viewModel => viewModel.VariantsCurrencies))
         {
             // Remove any content or the variants would be merged and not be cleared
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Drivers/PriceVariantsPartDisplayDriver.cs:57]: This comment is giving me a headache. Can you rewrite it?

File: src/Modules/OrchardCore.Commerce/Drivers/PriceVariantsPartDisplayDriver.cs [hunk 2]
```diff
@@ -16,13 +18,16 @@ public class PriceVariantsPartDisplayDriver : ContentPartDisplayDriver<PriceVari
 {
     private readonly IMoneyService _moneyService;
     private readonly IPredefinedValuesProductAttributeService _predefinedValuesProductAttributeService;
+    private readonly IOptions<CommerceSettings> _options;
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Drivers/PriceVariantsPartDisplayDriver.cs:21]: Please give it a more specific name so we don't need to scroll up here to figure out what kind of options it is.

File: src/Modules/OrchardCore.Commerce/Recipes/OrchardCore.Commerce.Development.Setup.recipe.json [hunk 3]
```diff
@@ -58,6 +58,9 @@
         "OrchardCore.Templates",
         "OrchardCore.Widgets",
 
+        "OrchardCore.Commerce",
+        "OrchardCore.Commerce.SessionCartStorage",
+        "OrchardCore.Commerce.CommerceSettingsCurrencySelector",
         // Themes
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Recipes/OrchardCore.Commerce.Development.Setup.recipe.json:64]: Add a line between.

File: test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs [hunk 4]
```diff
@@ -19,7 +19,7 @@ public static async Task<Uri> RunSetupAsync(UITestContext context)
                 SiteTimeZoneValue = "Europe/London",
             });
 
-        context.Exists(By.Id("mainNav"));
+        context.Driver.Title.ShouldBe("Orchard Setup");
```
Comment (DAud-IcI) [inline test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs:22]: Why is this necessary? I think the `context.Exists(By.Id("mainNav"))` is meant to check that the setup is concluded, while `Title.ShouldBe("Orchard Setup")` suggests the opposite.

File: test/OrchardCore.Commerce.Tests.UI/Tests/PriceVariantsPartTests/PersistencePriceVariantsTests.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            {
+                await context.SignInDirectlyAsync();
+                await context.CreateNewContentItemAsync("PriceVariantsProduct");
+
+                const string sku = "UITESTSKU";
+                const string price = "9999";
+
+                await context.ClickAndFillInWithRetriesAsync(By.Id("ProductPart_Sku"), sku);
+                await context.ClickAndFillInWithRetriesAsync(By.Id("PriceVariantsPart_VariantsValues__"), price);
+                await context.SetDropdownByTextAsync("PriceVariantsPart_VariantsCurrencies__", "USD");
```
Comment (DAud-IcI) [inline test/OrchardCore.Commerce.Tests.UI/Tests/PriceVariantsPartTests/PersistencePriceVariantsTests.cs:33]: Isn't it USD by default anyway?

File: test/OrchardCore.Commerce.Tests.UI/Tests/PriceVariantsPartTests/PersistencePriceVariantsTests.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+                await context.ClickReliablyOnSubmitAsync();
+                context.ShouldBeSuccess();
+
+                await context.GoToContentItemListAsync("PriceVariantsProduct");
+                await context.ClickReliablyOnAsync(By.XPath("//a[. = 'Edit']"));
+
+                context.Get(By.Id("ProductPart_Sku")).GetDomProperty("value").ShouldBe(sku);
+                context.Get(By.Id("PriceVariantsPart_VariantsValues__")).GetDomProperty("value").ShouldBe("9999.00");
```
Comment (DAud-IcI) [inline test/OrchardCore.Commerce.Tests.UI/Tests/PriceVariantsPartTests/PersistencePriceVariantsTests.cs:42]: Verify that the currency persists too.

**Review by DAud-IcI — CHANGES_REQUESTED** [review]

File: src/Modules/OrchardCore.Commerce/Drivers/PriceVariantsPartDisplayDriver.cs (hunk 1 above)
Comment (Psichorex) [reply src/Modules/OrchardCore.Commerce/Drivers/PriceVariantsPartDisplayDriver.cs:57]: Your headache is understandable I couldn't really understand this either.

File: test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs (hunk 4 above)
Comment (Psichorex) [reply test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs:22]: Well it was failing with `context.Exists(By.Id("mainNav"))` which I couldn't resolve because it made no sense to me.
I checked the logs and when this SetupHelpers.RunSetupAsnyc() was called from the new UI test the following happened:
ExecuteTestAfterSetupAsync() called the setup and all default values were presented to the setup page and Finish button was click but as soon as the setup finished the UI test went nuts right into the SignInDirectlyAsync() and it never reached the HomePage after setup. It wanted to check mainNav existance on the Login Page which ended up failing which is understandable.

File: test/OrchardCore.Commerce.Tests.UI/Tests/PriceVariantsPartTests/PersistencePriceVariantsTests.cs (hunk 6 above)
Comment (Psichorex) [reply test/OrchardCore.Commerce.Tests.UI/Tests/PriceVariantsPartTests/PersistencePriceVariantsTests.cs:42]: Can you provide an idea how to verify that? It is an asp-dropdown and there is actually nothing that is indicating the selected value. <option> tags don't even receive at least a selected class when you select is.

File: test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs (hunk 4 above)
Comment (DAud-IcI) [reply test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs:22]: > which I couldn't resolve because it made no sense to me.

What do you mean? It worked until now and we use this formula in most other projects' UI tests too.

> but as soon as the setup finished the UI test went nuts right into the SignInDirectlyAsync()

You shouldn't be redirected to the login page, that's an error that needs to be fixed.

More importantly, as I wrote the current `context.Driver.Title.ShouldBe("Orchard Setup");` passes while the setup is still processing. So you no longer guarantee to wait until the setup finished. This will lead to timing errors down the line and it will be a hell to debug then.

File: test/OrchardCore.Commerce.Tests.UI/Tests/PriceVariantsPartTests/PersistencePriceVariantsTests.cs (hunk 6 above)
Comment (DAud-IcI) [reply test/OrchardCore.Commerce.Tests.UI/Tests/PriceVariantsPartTests/PersistencePriceVariantsTests.cs:42]: You can use `By.CssSelector`. The `option:checked` pseudo-class actually picks the selected option for you. You have to add `.OfAnyVisibility()` too.

File: test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs (hunk 4 above)
Comment (Psichorex) [reply test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs:22]: After spending hours trying to figure out what could cause this issue I failed.
If I set up the site manually it works perfectly. The WelcomePage appears there is mainNav in the inspection so everything works as they should.
When I am running the UI test however the same thing happens.


There are subsequent  screens taken by the UI test. You can see it never reached the Welcome Page but I assume that the Setup has finished correctly because the logs say so:

`2022-08-08 12:37:25.0097  INFO > Click "Finish Setup" button
2022-08-08 12:37:25.0098 TRACE - > Execute behavior ClicksUsingClickMethodAttribute against "Finish Setup" button
2022-08-08 12:37:31.2786 TRACE - < Execute behavior ClicksUsingClickMethodAttribute against "Finish Setup" button (6.268s)
2022-08-08 12:37:31.2787  INFO < Click "Finish Setup" button (6.268s)
2022-08-08 12:37:31.4158  INFO > Exists applied to: 
By.Id: mainNav
2022-08-08 12:37:41.4371  INFO < Exists applied to: 
By.Id: mainNav (10.021s) >> OpenQA.Selenium.NoSuchElementException: Unable to locate visible element:...
2022-08-08 12:37:41.4469 - An exception has occurred while interacting with the page https://localhost:9095/Login?ReturnUrl=%2F (Orchard Setup).
2022-08-08 12:37:41.5674 - The test failed with the following exception: OpenQA.Selenium.NoSuchElementException: Unable to locate visible element:
- By: id "mainNav"
- Search time: 10.019s
- Search options: {Visibility=Visible, Timeout=5s, RetryInterval=0.5s, IsSafely=False}`

As you can see in the logs after the Finish Setup button is clicked it gets right to the Login page at least that is what I think by the log.

Comment (Psichorex) [reply test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs:22]: By the way the same substitution is used in WTW and there also if I swap `context.Driver.Title.Shouldbe()` to `context.Exist()` WTW UI tests are starting to fail on unable to locate visible element "navbar"

Comment (DAud-IcI) [reply test/OrchardCore.Commerce.Tests.UI/Helpers/SetupHelpers.cs]: SNOW is different, because there _everything_ requires authentication. Anonymous access is intentionally disabled. When the setup is finished and you are sent to the home page, you are immediately redirected to the login screen. That's the expected behavior. OCC has no such restriction.
